zerver/tests/test_external.py

```diff
@@ -118,6 +129,35 @@ def test_hit_ratelimits(self) -> None:
 
             self.assert_json_success(result)
 
+    def test_hit_ratelimits_as_ip(self) -> None:
+        add_ratelimit_rule(1, 5, domain='api_by_ip')
+        try:
+            RateLimitedIPAddr("127.0.0.1").clear_history()
+
+            start_time = time.time()
+            for i in range(6):
+                with mock.patch('time.time', return_value=(start_time + i * 0.1)):
+                    result = self.send_unauthed_api_request()
+
+            self.assertEqual(result.status_code, 429)
+            json = result.json()
+            self.assertEqual(json.get("result"), "error")
+            self.assertIn("API usage exceeded rate limit", json.get("msg"))
+            self.assertEqual(json.get('retry-after'), 0.5)
+            self.assertTrue('Retry-After' in result)
+            self.assertEqual(result['Retry-After'], '0.5')
+
+            # We actually wait a second here, rather than force-clearing our history,
```

Comment:
I would argue that we're not "actually" waiting a second -- we're _pretending_ to wait a second.

Edit: I see that this is copy/pasted from above, and we fix this in the next commit.  Slight preference for mentioning this in the commit message, or doing the refactor in the subsequent commit first, so we don't propagate this mis-statement.